### PR TITLE
Do not watch the global cache directories

### PR DESCRIPTION
**Pull request by wolfs**
The caches directory is completely managed by Gradle, so it doesn't make sense to watch it and only adds some overhead.

Fixes https://github.com/gradle/gradle/issues/12209

### CODE REVIEW

**Comment by wolfs** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheScopeMapping.java [hunk 1]
```diff
@@ -42,7 +45,7 @@ public DefaultCacheScopeMapping(File userHomeDir, @Nullable File projectCacheDir
 
     @Override
     public File getBaseDirectory(@Nullable Object scope, String key, VersionStrategy versionStrategy) {
-        if (key.equalsIgnoreCase("projects") || key.equalsIgnoreCase("tasks") || !key.matches("\\p{Alpha}+[-//.\\w]*")) {
+        if (key.equalsIgnoreCase("projects") || key.equalsIgnoreCase("tasks") || !key.matches("\\p{Alpha}+[-/.\\w]*")) {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheScopeMapping.java:48]: Would make sense to extract that pattern into a constant.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java [hunk 2]
```diff
@@ -39,8 +39,8 @@ public File transform(File source, CompleteFileSystemLocationSnapshot sourceSnap
             // Directories are ok to use outside the cache
             return source;
         }
-        for (AdditiveCache cache : additiveCaches) {
-            for (File root : cache.getAdditiveCacheRoots()) {
+        for (GlobalCache cache : globalCaches) {
+            for (File root : cache.getGlobalCacheRoots()) {
                 if (source.toPath().startsWith(root.toPath())) {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java:44]: Why don't we inject `GlobalCacheLocations` here instead? All we need to know here is whether or not a particular path is in a global cache.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/GlobalCache.java [hunk 3]
```diff
@@ -20,15 +20,14 @@
...
+ * Represents a location for global Gradle caches.
  *
- * Over time more files can be added to an additive cache, but existing files
- * can never be modified. Files in additive caches can only be deleted as part
- * of cleanup.
+ * The global cache is managed by Gradle, so we Gradle needs to take care
+ * of informing all the infrastructure about changes to it.
  */
-public interface AdditiveCache {
+public interface GlobalCache {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/vfs/GlobalCache.java:28]: I don't think these should be in the `org.gradle.internal.vfs` package. We probably shouldn't keep anything in that package in `:core`, but these classes don't strike me as VFS-related at all.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java [hunk 4]
```diff
@@ -163,14 +163,14 @@ FileHasher createCachingFileHasher(StringInterner stringInterner, CrossBuildFile
...
-            AdditiveCacheLocations additiveCacheLocations,
+            GlobalCacheLocations globalCacheLocations,
             FileHasher hasher,
             FileSystem fileSystem,
             Stat stat,
             StringInterner stringInterner,
             ListenerManager listenerManager
         ) {
-            Predicate<String> watchFilter = path -> !additiveCacheLocations.isInsideAdditiveCache(path);
+            Predicate<String> watchFilter = path -> !globalCacheLocations.isInsideGlobalCache(path);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java:173]: Maybe we could comment here why global cache locations are not watched.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/DefaultGlobalCacheLocations.java [hunk 5]
```diff
@@ -22,21 +22,26 @@
...
         }
-        this.additiveCacheRoots = additiveCacheRoots;
+        this.globalCacheRoots = globalCacheRoots;
     }
 
     @Override
-    public boolean isInsideAdditiveCache(String path) {
-        return additiveCacheRoots.contains(path);
+    public boolean isInsideGlobalCache(String path) {
+        return globalCacheRoots.contains(path);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/vfs/DefaultGlobalCacheLocations.java:40]: I wonder how a `FileHierarchySet` based on `Path` objects would work performance wise.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java [hunk 6]
```diff
@@ -82,7 +88,7 @@ private static File validateReadOnlyCache(DocumentationRegistry documentationReg
                 CacheLayout.ROOT.getKey() + " directory at " + root + " . Please follow the instructions at " + docLink);
             return null;
         }
-        return cacheDir;
+        return cacheDir.getAbsoluteFile();
```
Comment (lptr) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java:91]: Why is this needed now?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java [hunk 7]
```diff
@@ -56,18 +58,22 @@ public DefaultArtifactCaches(CacheScopeMapping cacheScopeMapping,
             IncubationLogger.incubatingFeatureUsed("Shared read-only dependency cache");
             File baseDir = validateReadOnlyCache(documentationRegistry, new File(roCache));
             if (baseDir != null) {
+                readOnlyCacheBaseDir = baseDir;
                 readOnlyCacheMetadata = new DefaultArtifactCacheMetadata(cacheScopeMapping, baseDir);
```
Comment (lptr) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java:62]: This code here is pretty bad, using `null`s to tie things together. I wonder if we could somehow not make it worse, though. It seems that `DefaultArtifactCacheMetadata` should be the one implementing the `GlobalCache` interface, no?

File: subprojects/files/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java [hunk 8]
```diff
@@ -218,14 +254,10 @@ boolean contains(String filePath, int offset) {
...
-            } else {
-                prefixes.add(depth + ":" + prefix.replace(File.separatorChar, '/'));
-            }
+        public void collect(int depth, HierarchyConsumer consumer) {
+            consumer.consumePrefix(depth, prefix);
             for (Node child : children) {
-                child.collect(depth + 1, prefixes);
+                child.collect(depth + 1, consumer);
             }
         }
```
Comment (lptr) [inline subprojects/files/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java:262]: Both `collect` and `consume` sounds a bit weird to me. This sounds like a visitor pattern to me. Shall we call it that?

File: subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java [hunk 9]
```diff
@@ -35,7 +35,7 @@ public CacheRepositoryServices(File gradleUserHomeDir, @Nullable File projectCac
         this.projectCacheDir = projectCacheDir;
     }
 
-    protected CacheScopeMapping createCacheScopeMapping() {
+    protected DefaultCacheScopeMapping createCacheScopeMapping() {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java:38]: `CacheScopeMapping` does not implement `GlobalCache`, which `DefaultCacheScopeMapping` does. Our injection framework only seems to detect interfaces implemented by the type returned from the creation method, so we need to return `DefaultCacheScopeMapping` here so that it is picked up when injecting `List<GlobalCache>`.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java (hunk 6 above)
Comment (wolfs) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java:91]: I want `cacheDir` to be an absolute path. Currently, it is read from whatever the environment variable is. That is why I think it is a good idea to convert it into an absolute path here.

File: subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java:38]: That is weird. Shouldn't we fix that problem?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java (hunk 6 above)
Comment (lptr) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java:91]: Why not do this where we call `validateReadOnlyCache()` from?

**Comment by lptr** [comment]
Should we change the title of the PR, given there are multiple directories we are now not watching?

**Comment by wolfs** [comment]
I don't know what we could do to see if we miss some global caches. Do you have any ideas?

**Comment by wolfs** [comment]
I renamed the PR to reflect that we are not watching the global cache directories.

File: subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java:38]: Ah, I think I understand why the interfaces need to be part of the return type: They need to be known at compilation time, since the actual creation method will not be invoked when it doesn't provide any required service. So I don't see a way how to fix this.

**Comment by lptr** [comment]
> I don't know what we could do to see if we miss some global caches. Do you have any ideas?

No real ideas, but it seems we could somehow make the "global cachiness" of a thing more integral? Like expose some methods on `GlobalCache` that you'd get the use of when you implement a global cache. Again, no concrete ideas.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java]: Doing that now.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/SplitFileHasher.java [hunk 10]
```diff
@@ -29,17 +29,17 @@
 public class SplitFileHasher implements FileHasher {
     private final FileHasher globalHasher;
     private final FileHasher localHasher;
-    private final AdditiveCacheLocations additiveCacheLocations;
+    private final GlobalCacheLocations additiveCacheLocations;
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/SplitFileHasher.java]: Done.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/SplitFileHasher.java [hunk 11]
```diff
@@ -29,17 +29,17 @@
...
 
-    public SplitFileHasher(FileHasher globalHasher, FileHasher localHasher, AdditiveCacheLocations additiveCacheLocations) {
+    public SplitFileHasher(FileHasher globalHasher, FileHasher localHasher, GlobalCacheLocations additiveCacheLocations) {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/SplitFileHasher.java]: Done.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/SplitResourceSnapshotterCacheService.java [hunk 12]
```diff
@@ -27,17 +27,17 @@
 public class SplitResourceSnapshotterCacheService implements ResourceSnapshotterCacheService {
     private final ResourceSnapshotterCacheService globalCache;
     private final ResourceSnapshotterCacheService localCache;
-    private final AdditiveCacheLocations additiveCacheLocations;
+    private final GlobalCacheLocations additiveCacheLocations;
 
-    public SplitResourceSnapshotterCacheService(ResourceSnapshotterCacheService globalCache, ResourceSnapshotterCacheService localCache, AdditiveCacheLocations additiveCacheLocations) {
+    public SplitResourceSnapshotterCacheService(ResourceSnapshotterCacheService globalCache, ResourceSnapshotterCacheService localCache, GlobalCacheLocations additiveCacheLocations) {
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/SplitResourceSnapshotterCacheService.java]: Done.

File: subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java:38]: I think I would keep the `GlobalCache` thing from the interface, since it seems to be something internal which shouldn't be used by clients of `CacheScopeMapping`.

File: subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheScopeMapping.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCacheScopeMapping.java]: Extracted.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/GlobalCache.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/GlobalCache.java]: Moved to `org.gradle.cache` in `persistentCache`.

File: subprojects/core/src/main/java/org/gradle/internal/vfs/DefaultGlobalCacheLocations.java (hunk 5 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/vfs/DefaultGlobalCacheLocations.java]: Do you want to add an issue somewhere?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java (hunk 6 above)
Comment (wolfs) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/DefaultArtifactCaches.java]: Doing the `absoluteFile` now outside of `validateReadOnlyCache`.

File: subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java:38]: I still don't like how error-prone this is. We should at least add a comment here saying why we are deviating from the typical way of returning the interface.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java [hunk 13]
```diff
@@ -39,13 +38,9 @@ public File transform(File source, CompleteFileSystemLocationSnapshot sourceSnap
...
         }
-        for (AdditiveCache cache : additiveCaches) {
-            for (File root : cache.getAdditiveCacheRoots()) {
-                if (source.toPath().startsWith(root.toPath())) {
-                    // It's not expected that the file will be deleted, so assume it is ok to use outside the cache
-                    return source;
-                }
-            }
+        if (globalCacheLocations.isInsideGlobalCache(source.getAbsolutePath())) {
+            // It's not expected that the file will be deleted, so assume it is ok to use outside the cache
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java:42]: When the code was talking about _additive(-only)_ caches, it made sense to derive this conclusion. Now it talks about global caches, and the conclusion doesn't seem logical anymore. I wonder how this could be clarified. Maybe just state that global caches are additive-only?

File: subprojects/files/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java (hunk 8 above)
Comment (lptr) [reply subprojects/files/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java:262]: The parameter name should be `visitor`.

File: subprojects/launcher/launcher.gradle.kts [hunk 14]
```diff
@@ -77,8 +77,10 @@ dependencies {
     integTestRuntimeOnly(files(toolsJar))
 }
 
-tasks.jar {
-    val classpath = listOf(":bootstrap", ":baseServices", ":coreApi", ":core").joinToString(" ") {
+evaluationDependsOn(":persistentCache")
```
Comment (lptr) [inline subprojects/launcher/launcher.gradle.kts:80]: Gah. Why is this needed?

Comment (wolfs) [reply subprojects/launcher/launcher.gradle.kts:80]: Because the `persistentCache` project is not configured at the time we access the `jar` task on it if I don't add it. I think the other tests work accidentally. I'll ask how we can handle this code without reaching into the other projects.

Comment (lptr) [reply subprojects/launcher/launcher.gradle.kts:80]: Can't we use a provider here to configure the classpath attribute that would be evaluated at execution time insted?

Comment (lptr) [reply subprojects/launcher/launcher.gradle.kts:80]: We should probably also use `locationOnly` instead of `get()`, no?

Comment (wolfs) [reply subprojects/launcher/launcher.gradle.kts]: I am now using a new configuration to configure the classpath. `locationOnly` only works on `FileSystemLocationProperty` and not on the providers I have here.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java (hunk 13 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/CopyingClasspathFileTransformer.java]: I changed the comment.

File: subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java:39]: I added a comment.

File: subprojects/files/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java (hunk 8 above)
Comment (wolfs) [reply subprojects/files/src/main/java/org/gradle/internal/file/DefaultFileHierarchySet.java:262]: Done.

**Comment by wolfs** [comment]
@lptr PTAL!
